src/prefect/agent.py

```diff
@@ -66,19 +57,42 @@ def __init__(
             self.default_infrastructure = Process()
             self.default_infrastructure_document_id = None
 
-    async def work_queue_id_from_name(self) -> Optional[UUID]:
+    async def get_work_queues(self) -> Iterator[WorkQueue]:
         """
-        For agents that were provided a work_queue_name, rather than a work_queue_id,
-        this function will retrieve the work queue ID corresponding to that name.
-        If no matching queue is found, a warning is logged and `None` is returned.
+        Loads the work queue objects corresponding to the agent's target work
+        queues. If any of them don't exist, they are created.
         """
-        if not self.work_queue_name:
-            raise ValueError("No work queue name provided.")
-        try:
-            work_queue = await self.client.read_work_queue_by_name(self.work_queue_name)
-            return work_queue.id
-        except ObjectNotFound:
-            return await self.client.create_work_queue(name=self.work_queue_name)
+
+        # if queues were cached less than 30 seconds ago, yield the cached values
+        now = pendulum.now("UTC")
+        if self._cache.get("ts", now.subtract(seconds=31)) > now.subtract(seconds=30):
+            for queue in self._cache["queues"]:
+                yield queue
+            return
+
+        # otherwise clear the cache and reload the work queues
+        self._cache = dict(ts=now, queues=[])
+
+        for name in self.work_queues:
+            try:
+                work_queue = await self.client.read_work_queue_by_name(name)
+            except ObjectNotFound:
+
+                # if the work queue wasn't found, create it
+                try:
+                    work_queue = await self.client.create_work_queue(name=name)
+                    self.logger.info(f"Created work queue '{name}'.")
+
+                # if creating it raises an exception, it was probably just
+                # created by some other agent; rather than entering a re-read
+                # loop with new error handling, we log the exception and
+                # continue.
+                except Exception as exc:
+                    self.logger.exception(exc)
```

Comment:
We'll already display the full traceback here so let's provide context as the string instead
```suggestion
                    self.logger.exception(f"Failed to create work queue {name!r}.")
```